### ISSUE TITLE
Bug: Various valid selector types incorrectly reporting as unknown type selectors

### ISSUE THREAD

**Issue by ntwb** [stylelint/stylelint#1352]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Bug: Various valid selector types incorrectly reporting as unknown type selectors:

• `:lang(x)`
• `:nth-child(x)`, `:nth-last-child(x)`, `:nth-last-of-type(x)`, `:nth-of-type(x)` parenthesis values
• `acronym`, `applet`, `big`, `hgroup`, `strike`, `submit`, `tt`

> Which rule, if any, is this issue related to?

 `selector-type-no-unknown`

> What CSS is needed to reproduce this issue?

#### `:lang`:
```css
:lang(X),
:lang(he-il) { 
  color: pink;
}
```

> Via https://css-tricks.com/almanac/selectors/l/lang/
> > _"The matching of X against the element's language value is performed case-insensitively for characters within the ASCII range. The identifier X does not have to be a valid language name."_

#### `nth-*`:
```css
.one li:nth-child(1)
.two :nth-last-child(-n+2) :nth-last-child(1),
.three :nth-of-type(3n+1),
.four:nth-of-type(even) {
  color: pink;
}
```

#### General: `acronym`, `applet`, `big`, `hgroup`, `strike`, `submit`, `tt`
```css

applet,
acronym,
big,
hgroup,
strike,
submit,
tt{
  color: pink;
}
```
> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "selector-type-no-unknown": true
  }
}
```

> Which version of stylelint are you using?

 `6.5.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint --config .stylelintrc test.css`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

Using the following consolidated CSS:
```css
:lang(x),
:lang(he-il),
.one li:nth-child(1)
.two :nth-last-child(-n+2) :nth-last-child(1),
.three :nth-of-type(3n+1),
.four:nth-of-type(even),
applet,
acronym,
big,
hgroup,
strike,
submit,
tt{
  color: pink;
}
```

The following warnings were flagged:
```shell
11:7 Unexpected unknown type selector "x" (selector-type-no-unknown) [stylelint]
12:7 Unexpected unknown type selector "he-il" (selector-type-no-unknown) [stylelint]
14:22 Unexpected unknown type selector "-n" (selector-type-no-unknown) [stylelint]
14:25 Unexpected unknown type selector "2" (selector-type-no-unknown) [stylelint]
14:44 Unexpected unknown type selector "1" (selector-type-no-unknown) [stylelint]
15:21 Unexpected unknown type selector "3n" (selector-type-no-unknown) [stylelint]
15:24 Unexpected unknown type selector "1" (selector-type-no-unknown) [stylelint]
16:19 Unexpected unknown type selector "even" (selector-type-no-unknown) [stylelint]
17:1 Unexpected unknown type selector "applet" (selector-type-no-unknown) [stylelint]
18:1 Unexpected unknown type selector "acronym" (selector-type-no-unknown) [stylelint]
19:1 Unexpected unknown type selector "big" (selector-type-no-unknown) [stylelint]
20:1 Unexpected unknown type selector "hgroup" (selector-type-no-unknown) [stylelint]
21:1 Unexpected unknown type selector "strike" (selector-type-no-unknown) [stylelint]
22:1 Unexpected unknown type selector "submit" (selector-type-no-unknown) [stylelint]
23:1 Unexpected unknown type selector "tt" (selector-type-no-unknown) [stylelint]
```

**Comment by davidtheclark** [comment]
I'd bet this is a straightforward fix. Here are my guesses:

 - With `:lang()` and `:nth-child()` it's probably because postcss-selector-parser treats the inside of the `:nth-child` function as a selector. So we need to always ignore selectors whose parent is `nth-child`.
- With the other keywords they probably just need to be added to a list. @ntwb are those all *obsolete* tags? If so, I wonder if we actually *do* want to warn about them, since they should be removed.

**Comment by jeddy3** [comment]
> With :lang() and :nth-child() it's probably because postcss-selector-parser treats the inside of the :nth-child function as a selector. So we need to always ignore selectors whose parent is nth-child.

I'm not sure why [`isStandardTypeSelector`](https://github.com/stylelint/stylelint/blob/master/src/utils/isStandardTypeSelector.js) isn't catching the `nth-child()` ones in this context. But yes, we'll definitely need to add a check for `:lang()` in there.

> are those all obsolete tags?

- I couldn't find `<submit>`.
- `<hgroup>` is "removed from the HTML5 (W3C) specification, but it still is in the WHATWG version of HTML." ([ref](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/hgroup))
- the rest are categorised as "obsolete".

> If so, I wonder if we actually do want to warn about them, since they should be removed.

Tough call. My guts says we should warn.

**Comment by ntwb** [comment]
> `<hgroup>` is "removed from the HTML5 (W3C) specification, but it still is in the WHATWG version of HTML." (ref)

I think this should _not_ cause a warning due to the continued widespread use of `hgroup`, as stated in the ref _"unlikely to go away"_, it is also still supported in [TinyMCE](https://github.com/tinymce/tinymce), there's [~1,000 plugins](https://github.com/search?q=user%3Awp-plugins+hgroup&type=Code) using `hgroup` to some degree, sadly I don't have available a way to grep the usage of `hgroup` for WordPress themes but I would expect its usage to be significantly higher than that of its use in plugins.

> I couldn't find `<submit>`.
> the rest are categorised as "obsolete".

Upon more investigation the instance of `<submit>` is in fact erroneously in the codebase and should be removed along with the rest of those "obsolete" elements/tags/keywords, hence a good pickup by stylelint 👍 

> @davidtheclark: @ntwb are those all obsolete tags? If so, I wonder if we actually do want to warn about them, since they should be removed.
> @jeddy3: Tough call. My guts says we should warn.

An enhancement to the existing `selector-type-no-unknown` rule that uses a modified warning for these, e.g  "Obsolete `keyword` type selector" warning?

**Comment by jeddy3** [comment]
FYI: https://github.com/stylelint/stylelint/pull/1357 is a pending PR to patch-up `isStandardTypeSelector`.

I've had a change of heart. All these tags, with the exception of `<submit>`, are *"known" in the specs* (even if they are labelled as obsolete etc)  and I think they shouldn't warning.

Otherwise this rule will *never be "finished"* i.e. we'll have to track the *HTML* specs and add breaking changes to this rule whenever a tag is labelled as obsolete. Something I think we should try to avoid :)

**Comment by ntwb** [comment]
> FYI: #1357 is a pending PR to patch-up isStandardTypeSelector.

👍 

> I've had a change of heart. All these tags, with the exception of `<submit>`, are "known" in the specs (even if they are labelled as obsolete etc) and I think they shouldn't warning.

Did you mean `<hgroup>` here rather than `<submit>`?

> Otherwise this rule will never be "finished" i.e. we'll have to track the HTML specs and add breaking changes to this rule whenever a tag is labelled as obsolete. Something I think we should try to avoid :)

Agreed 👍

**Comment by jeddy3** [comment]
> Did you mean `<hgroup>` here rather than `<submit>`?

Nope, `<submit>` isn't known in the specs. Whereas `<hgroup>` is. So, we should only warn for `<submit>`. I probably could of worded that sentence a little bit though :)

**Comment by davidtheclark** [comment]
All the above SGTM

**Comment by davidtheclark** [comment]
Closed by #1360

**Comment by ntwb** [comment]
Thanks 👍
